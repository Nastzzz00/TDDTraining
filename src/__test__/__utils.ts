import { ApolloServer } from "apollo-server";
import { typeDefs, resolvers } from "../graphql";

const constructTestServer = ({ context = {} }) => {
  const server = new ApolloServer({
    typeDefs,
    resolvers,
    context: session => context
  });

  return { server, context };
};

export { constructTestServer };
