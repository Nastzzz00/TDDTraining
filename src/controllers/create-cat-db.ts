import { IDBModel } from "../commons/types";
import { createUCCreateCat } from "../core/";

const createCreateCatDB = (db: IDBModel<any>) => async ca => {
  const createCat = createUCCreateCat();
  const newCat = createCat(ca);
  return db.insert(newCat);
};

export { createCreateCatDB };
