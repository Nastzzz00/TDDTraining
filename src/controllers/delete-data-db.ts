import { IDBModel } from "../commons/types";

const DeleteRecordByIDDB = (db: IDBModel<any>) => async id => {
  return db.deleteById(id);
};

export { DeleteRecordByIDDB };
