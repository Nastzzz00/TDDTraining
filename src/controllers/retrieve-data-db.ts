import { IDBModel } from "../commons/types";

const getAllDataDB = (db: IDBModel<any>) => async () => {
  return db.getAll();
};
const getByIDDB = (db: IDBModel<any>) => async id => {
  return db.getById(id);
};

export { getAllDataDB, getByIDDB };
