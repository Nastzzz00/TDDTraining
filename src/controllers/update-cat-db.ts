import { IDBModel } from "../commons/types";
import { updateUCCat } from "../core/_usecases";

const updateCatByIDDB = (db: IDBModel<any>) => async data => {
  const a = await db.getById(data.id);
  const updateScheduleLine = updateUCCat();
  const newScheduleLine = updateScheduleLine(data, a);
  return db.updateById({ ...newScheduleLine, id: data.id });
};

export { updateCatByIDDB };
