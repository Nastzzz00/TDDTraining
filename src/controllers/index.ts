import { getAllDataDB, getByIDDB } from "./retrieve-data-db";

//Cat
import { createCreateCatDB } from "./create-cat-db";
import { updateCatByIDDB } from "./update-cat-db";

export {
  //Cat
  getAllDataDB,
  getByIDDB,
  createCreateCatDB,
  updateCatByIDDB
};
