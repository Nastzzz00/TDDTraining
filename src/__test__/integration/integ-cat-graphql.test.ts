import { createTestClient } from "apollo-server-testing";
import gql from "graphql-tag";
import mockData from "../mocks/mock-data";
import * as controllers from "../../controllers";
import { constructTestServer } from "../__utils";

const {
  //Generic
  getAllDataDB,
  getByIDDB,

  //cat
  createCreateCatDB,
  updateCatByIDDB
} = controllers;

const userMock = {
  insert: jest.fn(async input => {
    return { id: "1", ...input };
  }),
  getById: jest.fn(async id => {
    const filterData = data => {
      if (data.id === id) {
        return data;
      }
    };
    const res = mockData.cat.filter(filterData);

    return res[0] || null;
  }),
  getAll: jest.fn(async () => {
    return mockData.cat;
  }),
  updateById: jest.fn(async input => {
    return { ...input };
  }),
  deleteById: async id => {
    const filterData = data => {
      if (data.id === id) {
        return data;
      }
    };
    const res = mockData.cat.filter(filterData);
    return res[0] || null;
  }
  // getAllBySupplierStatus: async id => {},
  // getAllByItem: async id => {},
  // getAllByScheduleLine: async id => {},
  // updateSupplierStatusItemById: async id => {},
  // updateAdminStatusPurchaseOrderById: async id => {},
};

const { server }: any = constructTestServer({
  context: {
    getAllCat: getAllDataDB(userMock),
    getCatById: getByIDDB(userMock),
    createCat: createCreateCatDB(userMock),
    updateCat: updateCatByIDDB(userMock)
  }
});

describe("Tests", () => {
  //Queries
  it("should fetch all cat", async () => {
    const CAT_ALL = gql`
      query {
        cats {
          id
          name
          lastname
          breed
        }
      }
    `;

    const { query } = createTestClient(server);
    const res = await query({ query: CAT_ALL });

    expect(res).toMatchSnapshot();
  });

  it("should fetch one cat", async () => {
    const SINGLE_CAT = gql`
      query u($id: String!) {
        getCat(id: $id) {
          id
          name
          lastname
          breed
        }
      }
    `;

    const { query } = createTestClient(server);
    const res = await query({
      query: SINGLE_CAT,
      variables: { id: "1" }
    });

    expect(res).toMatchSnapshot();
  });

  it("should error when no cat", async () => {
    const SINGLE_CAT = gql`
      query u($id: String!) {
        getCat(id: $id) {
          id
          name
          lastname
          breed
        }
      }
    `;

    const { query } = createTestClient(server);
    const res = await query({
      query: SINGLE_CAT,
      variables: { id: "" }
    });

    expect(res).toMatchSnapshot();
  });

  //Mutations

  it("should create a cat", async () => {
    const CREATE_CAT = gql`
      mutation createCat($cat: CreateCatInput!) {
        createCat(cat: $cat) {
          id
          name
          lastname
          breed
        }
      }
    `;

    const { mutate } = createTestClient(server);
    const res = await mutate({
      mutation: CREATE_CAT,
      variables: {
        cat: {
          name: "Jonas",
          lastname: "Nepomuceno",
          breed: "PERSIAN"
        }
      }
    });

    expect(res.errors).toBeUndefined();
    expect(userMock.insert.mock.calls.length).toBe(1);
    expect(res.data).toMatchObject({
      createCat: {
        name: "Jonas",
        lastname: "Nepomuceno",
        breed: "PERSIAN"
      }
    });
    expect(res).toMatchSnapshot();
  });

  // it('should delete a user', async () => {
  //   const DELETE_USER = gql`
  //     mutation u($id: ID!) {
  //       deleteUser(id: $id) {
  //         id
  //       }
  //     }
  //   `;
  //   const { mutate } = createTestClient(server);
  //   const res = await mutate({
  //     mutation: DELETE_USER,
  //     variables: { id: 'U1' },
  //   });
  //   expect(res.errors).toBeUndefined();
  //   expect(res).toMatchSnapshot();
  // });

  it("should update a cat", async () => {
    const UPDATE_CAT = gql`
      mutation u($cat: UpdateCatInput!) {
        updateCat(cat: $cat) {
          id
          name
          lastname
          breed
        }
      }
    `;

    const { mutate } = createTestClient(server);
    const res = await mutate({
      mutation: UPDATE_CAT,
      variables: {
        updateCat: {
          id: "1",
          name: "Jonas",
          lastname: "Nepomuceno",
          breed: "PUSAKAL"
        }
      }
    });

    expect(res).toMatchSnapshot();
  });
});
