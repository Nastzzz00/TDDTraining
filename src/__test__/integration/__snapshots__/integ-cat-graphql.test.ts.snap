// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests should create a cat 1`] = `
Object {
  "data": Object {
    "createCat": Object {
      "breed": "PERSIAN",
      "id": "1",
      "lastname": "Nepomuceno",
      "name": "Jonas",
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`Tests should error when no cat 1`] = `
Object {
  "data": undefined,
  "errors": Array [
    [ValidationError: Variable "$id" of type "String!" used in position expecting type "ID!".],
  ],
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`Tests should fetch all cat 1`] = `
Object {
  "data": null,
  "errors": Array [
    [GraphQLError: Cannot return null for non-nullable field Cat.id.],
  ],
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`Tests should fetch one cat 1`] = `
Object {
  "data": undefined,
  "errors": Array [
    [ValidationError: Variable "$id" of type "String!" used in position expecting type "ID!".],
  ],
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`Tests should update a cat 1`] = `
Object {
  "data": undefined,
  "errors": Array [
    [GraphQLError: Variable "$cat" of required type "UpdateCatInput!" was not provided.],
  ],
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;
